README.md

```diff
@@ -0,0 +1,486 @@
+rem(0.5);
+// ! TypeError: ‘rem’ expected a value of type Integer as its first argument; received 0.5
+
+rem(42, 0);
+// ! TypeError: ‘rem’ expected a value of type NonZeroInteger as its second argument; received 0
+```
+
+#### `UnaryType`
+
+`UnaryType` is used to construct types with one type variable. The `Array`
+type exported by sanctuary-def is defined via `UnaryType`.
+
+```javascript
+//    sum :: [Number] -> Number
+const sum =
+def('sum', {}, [$.Array($.Number), $.Number], xs => xs.reduce((x, y) => x + y, 0));
+
+sum([1, 2, 3, 4]);
+// => 10
+
+sum(['1', '2', '3', '4']);
+// ! TypeError: ‘sum’ expected a value of type (Array Number) as its first argument; received ["1", "2", "3", "4"]
+```
+
+To define a unary type `t a` one must provide:
+
+  - the name of `t`;
+
+  - a predicate which accepts any JavaScript value and returns `true` if
+    (and only if) the value is a member of `t x` for some type `x`;
+
+  - a function which takes any value of type `t a` and returns an array of
+    the values of type `a` contained in the `t`; and
+
+  - the type of `a`.
+
+```haskell
+UnaryType :: String -> (* -> Boolean) -> (t a -> [a]) -> Type -> Type
+```
+
+For example:
+
+```javascript
+//    Maybe :: Type -> Type
+const Maybe = $.UnaryType(
+  'sanctuary-def/Maybe',
+  x => x != null && x['@@type'] === 'sanctuary-def/Maybe',
+  maybe => maybe.isJust ? [maybe.value] : []
+);
+
+//    Nothing :: Maybe a
+const Nothing = {
+  '@@type': 'sanctuary-def/Maybe',
+  'isJust': false,
+  'isNothing': true,
+  'toString': () => 'Nothing',
+};
+
+//    Just :: a -> Maybe a
+const Just = x => ({
+  '@@type': 'sanctuary-def/Maybe',
+  'isJust': true,
+  'isNothing': false,
+  'toString': () => 'Just(' + JSON.stringify(x) + ')',
+  'value': x,
+});
+
+//    fromMaybe :: a -> Maybe a -> a
+const fromMaybe =
+def('fromMaybe', {}, [a, Maybe(a), a], (x, m) => m.isJust ? m.value : x);
+
+fromMaybe(0, Just(42));
+// => 42
+
+fromMaybe(0, Nothing);
+// => 0
+
+fromMaybe(0, Just('XXX'));
+// ! TypeError: ‘fromMaybe’ expected a value of type (Maybe Number) as its second argument; received Just("XXX")
+```
+
+#### `BinaryType`
+
+`BinaryType` is used to construct types with two type variables.
+
+To define a binary type `t a b` one must provide:
+
+  - the name of `t`;
+
+  - a predicate which accepts any JavaScript value and returns `true` if
+    (and only if) the value is a member of `t x y` for some types `x` and `y`;
+
+  - a function which takes any value of type `t a b` and returns an array of
+    the values of type `a` contained in the `t`;
+
+  - a function which takes any value of type `t a b` and returns an array of
+    the values of type `b` contained in the `t`;
+
+  - the type of `a`; and
+
+  - the type of `b`.
+
+```haskell
+BinaryType :: String -> (* -> Boolean) -> (t a b -> [a]) -> (t a b -> [b]) -> Type -> Type -> Type
+```
+
+For example:
+
+```javascript
+//    $Pair :: Type -> Type -> Type
+const $Pair = $.BinaryType(
+  'sanctuary-def/Pair',
```

Comment:
According to [`S.type`][1],

> `'@@type'` properties should use the form `'<package-name>/<type-name>'`, where `<package-name>` is the name of the npm package in which the type is defined.

It's not clear which "package name" we should use. On one hand this type *is* defined in the sanctuary-def package (though only in the README). On the other hand, using `'sanctuary-def/Pair'` here may lead readers to assume that they should use the `sanctuary-def` namespace in their own projects.

Your `'my-app/Pair'` suggestion would avoid this confusion. Perhaps `'my-package/Pair'` is best of all.


[1]: https://github.com/plaid/sanctuary/blob/22828c682bda3847c58274da8719b59d5d258e47/index.js#L339-L357